backport/datasync/data_metadata.py

```diff
@@ -0,0 +1,190 @@
+    sourceDescription = row.pop("sourceDescription")
+    nonRedistributable = row.pop("nonRedistributable")
+    displayJson = row.pop("display")
+
+    display = json.loads(displayJson)
+    partialSource = json.loads(sourceDescription)
+    variableMetadata = dict(
+        **_omit_nullable_values(variable),
+        type="mixed",  # precise type will be updated further down
+        nonRedistributable=bool(nonRedistributable),
+        display=display,
+        source=dict(
+            id=sourceId,
+            name=sourceName,
+            dataPublishedBy=partialSource.get("dataPublishedBy") or "",
+            dataPublisherSource=partialSource.get("dataPublisherSource") or "",
+            link=partialSource.get("link") or "",
+            retrievedDate=partialSource.get("retrievedDate") or "",
+            additionalInfo=partialSource.get("additionalInfo") or "",
+        ),
+    )
+
+    entityArray = (
+        variable_data[["entityId", "entityName", "entityCode"]]
+        .drop_duplicates(["entityId"])
+        .rename(columns={"entityId": "id", "entityName": "name", "entityCode": "code"})
+        .set_index("id", drop=False)
+        .to_dict(orient="records")
+    )
+
+    yearArray = (
+        variable_data[["year"]]
+        .drop_duplicates(["year"])
+        .rename(columns={"year": "id"})
+        .set_index("id", drop=False)
+        .to_dict(orient="records")
+    )
+
+    variableData = variable_data[["year", "entityId", "value"]].rename(
+        columns={"year": "years", "entityId": "entities", "value": "values"}
+    )
+
+    variableMetadata["type"] = _infer_variable_type(variableData["values"])
+
+    variableMetadata["dimensions"] = {
+        "years": {"values": yearArray},
+        "entities": {"values": entityArray},
+    }
+
+    # convert timestamp to string
+    time_format = "%Y-%m-%dT%H:%M:%S.000Z"
+    for col in ("createdAt", "updatedAt"):
+        variableMetadata[col] = variableMetadata[col].strftime(time_format)  # type: ignore
+
+    return variableMetadata
+
+
+def _infer_variable_type(values: pd.Series) -> str:
+    try:
+        values = pd.to_numeric(values)
+        inferred_type = pd.api.types.infer_dtype(values)
+        if inferred_type == "floating":
+            return "float"
+        elif inferred_type == "integer":
+            return "int"
+        else:
+            raise NotImplementedError()
+    except ValueError:
+        if values.map(lambda s: isinstance(s, str)).all():
+            return "string"
+        else:
+            return "mixed"
+
+
+def _convert_to_numeric(values: pd.Series) -> pd.Series:
+    values = values.astype(str).replace("nan", np.nan)
+    # raises ValueError if any value is not numeric or float
+    try:
+        return values.map(int)
+    except ValueError:
+        pass
+    # perhaps they're all floats
+    try:
+        return values.map(float)
+    except ValueError:
+        pass
+    # otherwise return them as strings
```

Comment:
This function doesn't look very robust, but if you have tested it, I hope it works in practice.

Reply:
Yeah, it sucks. I copied it from grapher codebase. I'll add some unit tests and make it clear what it does.